lib/network_monitor.c

```diff
@@ -172,42 +178,48 @@ mptcpd_in_addr_compare(void const *a, void const *b, void *user_data)
  * @see l_queue_find()
  * @see l_queue_remove_if()
  */
-static bool mptcpd_in_addr_match(void const *a, void const *b)
+static bool mptcpd_sockaddr_match(void const *a, void const *b)
 {
-        struct mptcpd_in_addr  const *const lhs = a;
+        struct sockaddr        const *const lhs = a;
         struct mptcpd_rtm_addr const *const rhs = b;
 
         assert(lhs);
         assert(rhs);
-        assert(lhs->family == AF_INET || lhs->family == AF_INET6);
-
-        bool matched = lhs->family == rhs->ifa->ifa_family;
-
-        if (matched) {
-                if (lhs->family == AF_INET) {
-                        /*
-                          Kernel nla_put_in_addr() inserts a big
-                          endian 32 bit unsigned integer, not struct
-                          in_addr.
-                        */
-                        uint32_t const sa = *(uint32_t const*) rhs->addr;
-
-                        matched = (lhs->addr.addr4.s_addr == sa);
-                } else {
-                        struct in6_addr const *const sa =
-                                (struct in6_addr const*) rhs->addr;
-
-                        /**
-                         * @todo Is memcmp() suitable in this case?
-                         *       Do we need to worry about the
-                         *       existence of uninitialized bytes in
-                         *       the IPv6 address byte array.
-                         */
-                        matched = (memcmp(&lhs->addr.addr6.s6_addr,
-                                          sa,
-                                          sizeof(lhs->addr.addr6.s6_addr))
-                                   == 0);
-                }
+        assert(lhs->sa_family == AF_INET || lhs->sa_family == AF_INET6);
+
+        bool matched = (lhs->sa_family == rhs->ifa->ifa_family);
+
+        if (!matched)
+                return matched;
+
+        if (lhs->sa_family == AF_INET) {
+                struct sockaddr_in const *const addr =
+                        (struct sockaddr_in const *) lhs;
+
+                /*
+                  Kernel nla_put_in_addr() inserts a big endian 32 bit
+                  unsigned integer, not struct in_addr.
+                */
+                uint32_t const sa = *(uint32_t const*) rhs->addr;
+
+                matched = (addr->sin_addr.s_addr == sa);
+        } else {
+                struct sockaddr_in6 const *const addr =
+                        (struct sockaddr_in6 const *) lhs;
+
+                struct in6_addr const *const sa =
+                        (struct in6_addr const*) rhs->addr;
+
+                /**
+                 * @todo Is memcmp() suitable in this case?  Do we
+                 *       need to worry about the existence of
+                 *       uninitialized bytes in the IPv6 address byte
+                 *       array.
+                 */
```

Comment:
Should be safe. To be a little more paranoid you could reference the u6_addr8 array inside the nested struct/union